Ticket struct and related interfaces

yondonfu opened the pull request:
**What does this pull request do? Explain your changes. (required)**

Adds common PM ticket related code that will be used by both a sender and a recipient in the PM protocol.

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- Adds the `pm` package
- Adds a `Ticket` struct that reflects the data included in on-chain redemption transactions. Note: The auxiliary data for expiration validation is omitted for the time being until a concrete decision is made on our desired expiration policy
- Adds a `Broker` interface that serves as an abstraction over the on-chain smart contract and its functions
- Adds a `Validator` interface with functions for validating a ticket and checking for a winning ticket. The `BrokerValidator` implementation uses an implementation of the `Broker` interface to check for used tickets (replay prevention) and an implementation of the `SigVerifier` interface to validate signatures over ticket hashes
- Adds a `SigVerifier` interface with functions for verifying signatures. The `BrokerSigVerifier` implementation uses an implementation of the `Broker` interface to check for approved signers when verifying a signature.

**How did you test each of these updates (required)**

- Unit tests for each new component with stubbed out dependencies

**Does this pull request close any open issues?**
<!-- Fixes # -->

- Fixes #626 
- Fixes #628

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

yondonfu committed:
pm: Add pm package

Included Ticket data structure

yondonfu committed:
pm: Validator, SigVerifier and Broker interfaces

- Validator interface for validating tickets with a validator
implementation that relies on the Broker interface for used tickets
- SigVerifier interface for verifying ECDSA signatures with an
ApprovedSigVerifier implementation that relies on the Broker interface for
approved signers
- Broker interface for an abstraction of the on-chain smart contract

yondonfu committed:
pm: Validator fixes

Fixes signature verification and Solidity keccak256 hashing

j0sh commented on pm/ticket.go:
```diff
@@ -0,0 +1,109 @@
+	FaceValue *big.Int
+
+	// WinProb represents how likely a ticket will win
+	WinProb *big.Int
```
Just wondering, how were you planning on representing a probability as an integer?

yondonfu replied:
`WinProb` would just be an integer in the range 0 through 2^256 - 1 with the winning ticket check defined as `H(SIG(H(T)), recipientRand) < T.WinProb)` (the hash result is also in the range 0 through 2^256 - 1). Since `MaxWinProb = 2^256 - 1`, each `WinProb` value is can be thought of as a probability represented as the fraction `WinProb / MaxWinProb`.

eladmallel commented on pm/sigverifier.go:
```diff
@@ -0,0 +1,58 @@
+		return true, nil
+	}
+
+	approved, err := sv.broker.IsApprovedSigner(addr, rec)
```
I'm wondering about the performance implications this will have. Under the following conditions I think this might be problematic:

- If we assume a naive implementation, this would result in a function call to an ETH node every time, which could take a while
- If ticket validation is a blocker for processing a segment

Will these conditions be true when we implement `Broker` and integrate ticket validation in the `ServeSegment` flow?

yondonfu replied:
Great point. Here is a summary of an offline discussion about this:

An ETH call to check for approved signers to validate every ticket would create a performance hit for ticket validation which affects segment processing performance. In the short term for MVS, we can remove the approved signer feature client side and instead the recipient will just check that the signature corresponds to the sender address in the ticket (only requires client side cryptographic operations). In the longer term, we could implement an caching layer that replicates state from the on-chain contract so that a recipient can access a sender's approved signers without an ETH call.

j0sh replied:
>  In the longer term, we could implement either an caching layer

One way is to to lookup the list when the first segment of a stream comes in. That list only needs to be retained for the duration of the stream (eg, transcode loop). There'd be a hit for the very first segment, but subsequent lookups can remain local.

I wonder if borrowing the Explorer's GraphQL caching layer would also help with the initial lookup.

eladmallel replied:
The Graph is definitely worth exploring for this use case. We just need to make sure it has sufficiently efficient response times and no hangups, since we don't want The Graph to be the source of segment processing delays :)

yondonfu replied:
Interesting approach. O would likely also want to monitor revocation events for the duration of the stream in order to determine when certain previously approved signers will be considered revoked (and thus their signatures deemed invalid if submitted on-chain). 

Tracking this and other possible areas of exploration in #645

eladmallel commented on pm/sigverifier_test.go:
```diff
@@ -0,0 +1,76 @@
+		t.Fatal(err)
+	}
+
+	signer := crypto.PubkeyToAddress(signerPrivKey.PublicKey)
```
Perhaps rename `signer` to `unapprovedSigner` to increase readability? - I noticed we later approve `signer`, but perhaps then we can just create a new signer instance and name that one `approvedSigner`?

eladmallel commented on pm/sigverifier_test.go:
```diff
@@ -0,0 +1,76 @@
+	signerSig, err := crypto.Sign(personalHash, signerPrivKey)
+
+	b := &stubBroker{
+		usedTickets:     make(map[ethcommon.Hash]bool),
```
Why do we need to instantiate `usedTickets` for this test?

yondonfu replied:
I've replaced declarations of `stubBroker` struct literals with a call to the constructor `newStubBroker` which creates a `stubBroker` with all fields initialized and ready for use.

My thought was to try to avoid any accidental usage of `stubBroker` that leads to accessing a nil map.

eladmallel commented on pm/sigverifier_test.go:
```diff
@@ -0,0 +1,76 @@
+	// Test invalid signature for non-approved signer
+	valid, err := sv.Verify(sender, signerSig, msg)
+	if err != nil {
+		t.Fatal(err)
```
I think that starting from this line forward you want to switch from `t.Fatal` to `t.Fail`, since I believe `Fatal` stops the test function on that line and does not proceed, versus `Fail` that allows the test to continue, therefore not masking the rest of the checks in this function.

wdyt?

j0sh replied:
We also use `t.Error` elsewhere in the codebase to indicate "fail but continue"

eladmallel replied:
So what should be the consistent way to do it? `t.Error` or `t.Fail`?

yondonfu replied:
`t.Error` logs the provided message and then triggers a `t.Fail` https://golang.org/pkg/testing/#T.Error while `t.Fail` just marks the function as having failed without a message logged

eladmallel replied:
Sounds like `t.Error` is the choice then!

eladmallel commented on pm/ticket.go:
```diff
@@ -0,0 +1,109 @@
+// Broker is an interface which serves as an abstraction over an on-chain
+// smart contract that handles the administrative tasks in a probabilistic micropayment protocol
+// including processing deposits and pay outs
+type Broker interface {
```
What are the pros/cons of having the `Broker` interface in the `ticket.go` file, versus creating a new file like `broker.go` ?

yondonfu replied:
I didn't have a strong reason for keeping the `Broker` interface in `pm/ticket.go`. Since there is not an concrete implementation for this interface yet, I didn't bother putting it into it's own file. I can move this interface into it's own file thought - it seems cleaner and will probably be helpful later on for the concrete implementation

eladmallel commented on pm/validator.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// ValidateTicket checks if a ticket is valid
+func (v *validator) ValidateTicket(ticket *Ticket, sig []byte, recipientRand *big.Int) error {
```
Wondering when/where should we introduce the HMAC-based validation?

yondonfu replied:
My plan was to introduce HMAC validation as a part of https://github.com/livepeer/go-livepeer/issues/630. The motivation behind this was to have the `Validator` object be responsible for validating a ticket given a provided `recipientRand` value - this value could be randomly generated, could be generated using a HMAC based construction, etc. The HMAC validation logic would then be held in the recipient which passes the `recipientRand` value generated into the `Validator` object. Let me know if you have any thoughts/concerns on this separation of concerns!

j0sh replied:
At this point we assume that the given `recipientRand` is valid, and that it's been checked earlier in the flow? Makes sense.

Technically we could also keep the validation contained by making some `ReceipentRandValidator` interface accessible to the validator object, perhaps by passing it in during construction of the `validator` struct. No real preference though; incorporating it in #630 works too.

eladmallel commented on pm/validator.go:
```diff
@@ -0,0 +1,79 @@
+		return fmt.Errorf("invalid preimage provided for hash commitment recipientRandHash")
+	}
+
+	used, err := v.broker.IsUsedTicket(ticket)
```
Is `IsUsedTicket` focused on checking for random number replays, winning ticket replays, or both?

Intuitively seems to me that `Broker` should not be aware of local replay protection like `senderNonce` state, so based on your answer to the question above, would love to get a clearer sense of how we're organizing all the validations we need to run on each incoming ticket.

wdyt?

yondonfu replied:
`IsUsedTicket` checks if the broker has already processed this particular winning ticket (i.e. query the smart contract). Here is a summary of an offline discussion about this:

The broker keeps track of used winning tickets in order to protect a sender from a malicious recipient that tries to replay a winning ticket over and over again. Since the validator is the recipient, we do not need to protect against this scenario. However, we do need to protect against the scenario in which a sender purposefully sends used winning tickets to the recipient which results in the recipient being unable to redeem them. 

If we assume that the recipient also:

- Handles replay prevention for invalid already winning `recipientRand` values
- Uses an HMAC based on a secret generated on start up and the sender's address to validate `recipientRand` values

then the recipient does not need to call `IsUsedTicket` when validating a ticket.

The recipient will not accept tickets with an old winning `recipientRand` which means no old winning tickets created by the current sender will be accepted. Sender A can try to use winning tickets sent by sender B, but these tickets will not pass the HMAC validation step since the tickets include a different sender. If the recipient restarts, thereby clearing its current set of invalid already winning `recipientRand` values, it will generate a new secret so if a sender tries to replay an old `recipientRand` value, it will fail the HMAC validation which uses the recipient's new secret as the key.

Thus, the recipient will reject old winning tickets obviating the need for the call to `IsUsedTicket` which is a plus because it removes an ETH call from the ticket validation process which could've been a performance bottleneck.

eladmallel commented on pm/validator.go:
```diff
@@ -0,0 +1,79 @@
+// Note: This method does not check if a ticket is valid which is done using IsValidTicket
+// A ticket wins if:
+// H(SIG(H(T)), T.RecipientRand) < T.WinProb
+func (v *validator) IsWinningTicket(ticket *Ticket, sig []byte, recipientRand *big.Int) bool {
```
Just want to double check our approach regarding winning ticket replay protection: seems like that's not something we should worry about in the client because we invalidate `recipiendRand` values every time we submit a ticket, so there's no way an attempt to replay a winning ticket on O will actually work. It will be blocked because it will have a no-longer-valid `recipientRandHash` value.

Correct?

yondonfu replied:
Yep see my comment https://github.com/livepeer/go-livepeer/pull/638#discussion_r summarizing this conclusion.

eladmallel commented on pm/validator_test.go:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
+		t.Fatal("expected invalid recipient (non-null address)")
```
Change to `t.Fail` or `t.Error`? (writing this comment once for all occurrences in this test file)

eladmallel commented on pm/validator_test.go:
```diff
@@ -0,0 +1,167 @@
+		SenderNonce:       0,
+		RecipientRandHash: recipientRandHash,
+	}
+	// Set ticket as used in stub broker
```
So what makes this ticket used? - sounds like there should be a call to some stub function setting this ticket to used?

yondonfu replied:
This particular test has been removed in de637a5 since we no longer check the broker for used tickets during ticket validation

eladmallel commented on pm/validator_test.go:
```diff
@@ -0,0 +1,167 @@
+	// Set ticket as used in stub broker
+	b.RedeemWinningTicket(ticket, sig, recipientRand)
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
```
General concern: shouldn't we validate that the error we're getting is the specific error we expect?

eladmallel commented on pm/stub.go:
```diff
@@ -0,0 +1,137 @@
+	shouldVerify bool
+}
+
+func (sv *stubSigVerifier) SetShouldVerify(shouldVerify bool) {
```
I think this function name is a bit confusing. Perhaps changing it to `SetVerifyResult` is a bit clearer?

j0sh committed:
Fix orchestrator tests.

yondonfu committed:
pm: Validator performance and testing updates

- Remove unnecessary used ticket check which would've been a performance
bottleneck
- Improved validation error testing
- Misc code cleanup

eladmallel reviewed (approved):
🚢

j0sh reviewed (approved):
LGTM!

j0sh commented:
If this is being blocked because of CI, feel free to cherry pick https://github.com/livepeer/go-livepeer/pull/625/commits/85227b54c92d9202ad8e979532614ef88fcdd2cb and push
